spec/function.dd

```diff
@@ -173,124 +173,315 @@ int foo();
-    references of the result may be assumed to not refer to any object that
-    existed before the function call. For example:)
+$(H3 Strongly Pure Functions and Optimizations)
+
+$(P Strong purity enables optimizations based on their inability to read or modify data outside
+their parameters. In turn, that guarantees the function always returns the same result for the
+same arguments.)
+
+$(P A strongly pure function that accepts and returns by value can be assumed to return the same
+value for the same inputs. The compiler is allowed to memoize the result and reuse it for
+subsequent calls with identical arguments. Example:)
+
+---
+pure double fun(double);
+auto x = fun(1) + fun(1);
+// Code generated may be:
+// auto __t = fun(1);
+// auto x = __t + __t;
+---
+
+$(P For the purpose of deeming two calls to a strongly pure function equivalent, all arguments must
+be bitwise equal (as per the `is` operator), not logically equal (as per the `==` operator).
+Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+pure double fun(double x) { return x * 42; }
+pure int gun(int x) { return x ^ 42; }
+pure string hun(string s) { return s; }
+auto a = fun(1.0) + fun(1.0); // can memoize
+assert(-0.0 == 0.0 && -0.0 !is 0.0);
+auto b = fun(-0.0) + fun(0.0); // can NOT memoize, -0.0 and 0.0 have different bit patterns
+auto c = hun("ab") ~ hun("ab"); // can memoize ONLY if the two strings are interned
+---
+)
+
+$(P An implementation may not evaluate a strongly pure function speculatively. This is because a
+strongly pure function may abort the program, throw, or never terminate. Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+pure double fun(double x) { return x + 42; }
+pure double gun(size_t n)
+{
+    double result = 0;
+    foreach (i; 0 .. 100)
+    {
+        if (i == n) break;
+        result += fun(42);
+    }
+    return result;
+/*
+This code is NOT equivalent because it evaluates fun speculatively for n == 0:
+    double result = 0;
+    double __t = fun(42);
+    foreach (i; 0 .. 100)
+    {
+        if (i == n) break;
+        result += __t;
+    }
+    return result;
+*/
+}
+---
+)
+
+$(P A call to a `nothrow` strongly pure function that accepts and returns by value can be moved
+relative to its surrounding code, as long as its data dependencies are respected and there is no
+change in control flow. Example:)
+
+---
+pure double fun(double);
+pure double gun(double);
+void hun(double x)
+{
+    auto y = fun(x) + gun(x); // evaluate fun(x) and gun(x) in any order
+    foreach (i; 0 .. 100)
+    {
+        y += fun(10); // may hoist fun(10) outside the loop
+    }
+}
+---
+
+$(P If a strongly pure function throws, the assumptions related to memoization and references do
+not carry to the thrown object.)
+
+$(P Note: a strongly pure function may still have behavior inconsistent with memoization by e.g.
+using `cast`s or by changing behavior depending on the address of its parameters. An implementation
```

Comment:
I think you defined away memoization for equal `immutable` arguments with different addresses? (Not that I like it, but this seems to eliminate this concern.)

Reply:
Yes, that was the intent. Here's what I had in mind:

```d
pure string identity(string s) { return s; }
auto s1 = "abc",  s2 = s1.idup;
auto s3 = identity(s1), s4 = identity(s2);
```

If we allow s1 to be considered equivalent to s2, identity will always return s1.

Also: when in doubt, make it weakly pure.

Reply:
I see, you're saying I should eliminate that text. Will do.

Reply:
@tgehr will still keep ie because:

```
pure int* hmmm(int x) { return &x; }
```

Reply:
@tgehr simplified the text to, well, simplify

Reply:
Yes, my comment was purely about the parameter addresses. (By which I assumed you meant the actual address of references/pointers within the arguments, otherwise it should have said address of local variables.)